### PR TITLE
detect/sip: add sticky buffers to match headers v4

### PR DESCRIPTION
**Pull request by glongo**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x] I have read the contributing guide lines at
   https://docs.suricata.io/en/latest/devguide/contributing/contribution-process.html
- [x] I have signed the Open Information Security Foundation contribution agreement at
   https://suricata.io/about/contribution-agreement/ (note: this is only required once)
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)
- [x] I have created a ticket at
      https://redmine.openinfosecfoundation.org/projects/suricata/issues
      (if applicable)

Link to ticket: https://redmine.openinfosecfoundation.org/issues/
https://redmine.openinfosecfoundation.org/issues/6374

Describe changes:
- Store multiple values for a single header
- Header name in compact form is extended into long form

SV_BRANCH=https://github.com/OISF/suricata-verify/pull/1926

### CODE REVIEW

**Comment by catenacyber** [comment]
Thanks Giuseppe, quick remark : you can now have this in pure rust
cf https://github.com/OISF/suricata/pull/11316 for instance

**Comment by glongo** [comment]
> Thanks Giuseppe, quick remark : you can now have this in pure rust cf #11316 for instance

I saw your commit about rust keywords merged but I have a question:
Can multi-buffer keywords already be written in Rust?

**Comment by catenacyber** [comment]
> > Thanks Giuseppe, quick remark : you can now have this in pure rust cf #11316 for instance
> 
> I saw your commit about rust keywords merged but I have a question: Can multi-buffer keywords already be written in Rust?

I am doing it in #11316, this requires https://github.com/OISF/suricata/pull/11316/commits/7266b693a151fa9e991141d010a2129d2a65d7a1

**Review by catenacyber — CHANGES_REQUESTED** [review]
Thanks for the work :-)

This PR needs a rebase

- CI : 🟢 
- Code : checking now
- Commits segmentation : I would squash some together
- Commit messages : ok for me
- Git ID set : looks fine for me
- CLA : you already contributed
- Doc update : looks fine
- Redmine ticket : ok
- Rustfmt : ok
- Tests : SV ok
- Dependencies added: none

File: src/detect-sip-via.c [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+#define KEYWORD_ID       DETECT_AL_SIP_HEADER_VIA
+#define KEYWORD_TOSERVER 1
+#define KEYWORD_TOCLIENT 1
+
+#include "detect-sip-headers-stub.h"
+#include "detect-sip-via.h"
+
+void RegisterSipHeadersVia(void)
+{
+    DetectSipHeadersRegisterStub();
```
Comment (catenacyber) [inline src/detect-sip-via.c:38]: You can do them in rust now

File: rust/src/sip/parser.rs [hunk 2]
```diff
@@ -97,6 +97,22 @@ fn is_header_value(b: u8) -> bool {
     is_alphanumeric(b) || is_token_char(b) || b"\"#$&(),/;:<=>?@[]{}()^|~\\\t\n\r ".contains(&b)
 }
 
+fn expand_header_name(h: &str) -> &str {
+    match h {
+        "i" => "Call-ID",
```
Comment (catenacyber) [inline rust/src/sip/parser.rs:102]: Can we detect if some header uses the short or long form ?

Comment (glongo) [reply rust/src/sip/parser.rs:102]: Why should we detect that?

Comment (catenacyber) [reply rust/src/sip/parser.rs:102]: To get the most expressivity.
I do no say we should do that, I just asked if we can.

I guess this deserves some mention in the docs., like
```
This keyword matches on the From field that can be present in SIP headers.
It matches both short and regular form, and these forms cannot be distinguished by this keyword.
```

Comment (glongo) [reply rust/src/sip/parser.rs:102]: I'm not sure I understand what you mean. If a header uses the short form, it is 'expanded' ( converted to the long form) to enable matching in both cases with its related sticky buffer.
If this is what you are referring to, it has already been implemented.
 If not, I'm unclear on what you mean by 'detecting which form is used by a header.'

Comment (catenacyber) [reply rust/src/sip/parser.rs:102]: I meant 2 things :

> If a header uses the short form, it is 'expanded' ( converted to the long form) to enable matching in both cases with its related sticky buffer.

1. This should be in the docs

2. And the doc should also mention that there is only the normalized form, not the raw one (as we can have for http uri for instance where there are 2 keywords)

File: src/detect-sip-via.c (hunk 1 above)
Comment (glongo) [reply src/detect-sip-via.c:38]: Am I wrong, or has your PR for multibuffer keywords in rust not been merged yet?

Comment (catenacyber) [reply src/detect-sip-via.c:38]: It has been merged today :-)

File: rust/src/sip/detect.rs [hunk 3]
```diff
@@ -1,4 +1,4 @@
-/* Copyright (C) 2019 Open Information Security Foundation
+/* Copyright (C) 2024 Open Information Security Foundation
```
Comment (jufajardini) [inline rust/src/sip/detect.rs:1]: nit: when updating copyright years, we add the current year beside the initial one:
```suggestion
/* Copyright (C) 2019-2024 Open Information Security Foundation
```

**Review by jufajardini — CHANGES_REQUESTED** [review]
I only have a nit comment for the copyright year change in `sip/detect.rs`, but do agree with Philippe's comment that we should mention in the docs about the headers being normalized :)

File: rust/src/sip/parser.rs (hunk 2 above)
Comment (jufajardini) [reply rust/src/sip/parser.rs:102]: > I meant 2 things :
> 
> > If a header uses the short form, it is 'expanded' ( converted to the long form) to enable matching in both cases with its related sticky buffer.
> 
>     1. This should be in the docs
> 
>     2. And the doc should also mention that there is only the normalized form, not the raw one (as we can have for http uri for instance where there are 2 keywords)

+1

**Comment by glongo** [comment]
Replaced with https://github.com/OISF/suricata/pull/11620

**Pull request closed without merging**